docs/preview/features/message-pumps/customization.md

```diff
@@ -44,32 +80,43 @@ public class OrderMessagePump : MessagePump
 }
 ```
 
-## Filter messages based on message context
-
-When registering a new message handler, one can opt-in to add a filter on the message context which filters out messages that are not needed to be processed.
-
-This can be useful when you are sending different message types on the same queue. Another use-case is being able to handle different versions of the same message type which have different contracts because you are migrating your application.
+### Bring your own deserialization
 
-Following example shows how a message handler should only process a certain message when a property's in the context is present.
+You can also choose to extend the built-in message deserialization with additional deserializer to meet your needs. 
+This allows you to easily deserialize into different message formats or reuse existing (de)serialization capabilities that you already have without altering your message pump. 
 
-We'll use a simple message handler implementation:
+You start by implemeting an `IMessageBodySerializer`. The following example shows how an expected type can be transformed to something else. 
+The result type (in this case `OrderBatch`) will be then be used to check if there is an `IMessageHandler` registered with that message type.
 
 ```csharp
-public class OrderMessageHandler : IMessageHandler<Order>
+public class OrderBatchMessageBodySerializer : IMessageBodySerializer
 {
-    public async Task ProcessMessageAsync(Order order, MessageContext context, ...)
+    public async Task<MessageResult> DeserializeMessageAsync(string messageBody)
     {
-        // Do some processing...
-	}
+        var serializer = new XmlSerializer(typeof(Order[]));
+        using (var contents = new MemoryStream(Encoding.UTF8.GetBytes(messageBody)))
+        {
+            var orders = (Order[]) serializer.Deserialize(contents);
+            return MessageResult.Success(new OrderBatch(orders));
+        }
+    }
 }
 ```
 
-We would like that this handler only processed the message when the context contains `MessageType` equals `Order`.
+The registration of these message body handlers can be done just as easily as an `IMessageSerializer`:
 
 ```csharp
 public void ConfigureServices(IServiceCollection services)
 {
-    services.WithMessageHandler<OrderMessageHandler, Order>(context => context.Properties["MessageType"].ToString() == "Order");
+    // Register the message body serializer in the dependency container where the dependent services will be injected.
+    services.WithMessageHandler<OrderBatchMessageHandler>(..., messageBodySerializer: new OrderBatchMessageBodySerializer());
```

Comment:
Since it's just a new here, why not do this?
```suggestion
    services.WithMessageHandler<OrderBatchMessageHandler, OrderBatchMessageBodySerializer>(...);
```

Reply:
Because we don't have such an extension 😄 . We could, but it seemed at the time an unnecessary generic. I could add it, but it would mean we create arround 24 extra extensions with all the combinations. (plus tests of course).

Reply:
It feels more natural to me, but not worth the effort

Reply:
We could also change the existing ones so that there's a generic overload that creates it directly, and a implementation factory where you get the `IServiceProvider` to create your serializer.

Reply:
Can you give an example please?

Reply:
Now, we have these capabilities:
- `.WithMessageHandler(..., messageBodySerializer: new OrderSerializer());`
- `.WithmessageHandler(..., messageBodySerializerImplementationFactory: serviceProvider => new OrderSerializer());`

We could change it throughout all combinations with the generic instead:
- `.WithMessageHandler<..., OrderSerializer>(...);`
- `.WithmessageHandler(..., messageBodySerializerImplementationFactory: serviceProvider => new OrderSerializer());`

Reply:
I'm confused because isn't that what I suggested 🤔

Reply:
Well, I said we could _add_ also a new overload so that you could both send in the instance as the type and that would lead to 24+ extra extensions, or we could change the existing ones. Not sure what you suggested.

Reply:
Hmm, also, we're not gonna be able to do this approach for every overload bc it will result in some ambigous signatures (same amount of params, only with different generics, still fail).

Reply:
Let's just resolve this for now :)